components/style/attr.rs

```diff
@@ -603,3 +603,89 @@ pub struct AttrIdentifier {
     pub namespace: Namespace,
     pub prefix: Option<Prefix>,
 }
+
+/// Valid date string should be yyyy-mm-dd
+/// https://html.spec.whatwg.org/multipage/#valid-date-string
+pub fn is_valid_date_string(value: &str) -> bool {
+    // step 2
+    let value_vec: Vec<&str> = value.split('-').collect();
+    if !(value_vec.len() == 3) {
+        return false;
+    }
+    let year = value_vec[0];
+    let month = value_vec[1];
+    let day = value_vec[2];
+
+    // setp 1
+    let month_string = format!("{}-{}", year, month);
+    if !is_valid_month_string(&*month_string) {
```

Comment:
If it helps to split the responsibilities between `is_valid_month_string` and `is_valid_date_string` more cleanly, we can instead implement `parse a month component`, and have that algorithm return a `Result<(u32, u32), ParseError>`.

Reply:
Ah. I think this is a good way.